### PR TITLE
Generate keyword documentation with new GenerateDocumentation transformer

### LINKED ISSUE #311: [Transformer] Format documentation
From #300 

Transformer for formatting the documentation. By default it could use one style of the documentation (like Google docstring) but allow to switch to some other popular styles. 

Upon consideration: allowing to pass some kind of template that would be applied to documentation.

### PR DESCRIPTION
**Pull request by bhirsz**
Closes #311

We finally have transformer for generating the documentation! It may be useful to help creating resource file documentation. 

To allow providing custom templates as relative paths I had to abuse a little some of our internal code (like reading the configuration) - I've left comments and covered it with tests so it should be fine. If I will came with better solution, for example after some code refactor, I will change the implementation.

### CODE REVIEW

**Comment by bhirsz** [comment]
Robot Framework v4 fails because it doesn't support ``RETURN`` I have used in tests - I will update tests.

File: docs/source/transformers/GenerateDocumentation.rst [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+                Keyword
+                    [Documentation]
+                    ...    Arguments:
+                    ...        ${var} - :
+                    ...        ${var2} - 2:
+                    [Arguments]    ${var}    ${var2} = 2
+                    Step
+
+    Note that you can use Jinja templating features like if blocks. For example, if you want to put ``=`` between
+    argument name and default value only if default value is not empty, you can use::
```
Comment (mnojek) [inline docs/source/transformers/GenerateDocumentation.rst:242]:
```suggestion
    Note that you can use Jinja templating features like ``if`` blocks. For example, if you want to put ``=`` between
    the argument name and default value (only if default value is not empty), you can use::
```

File: docs/source/transformers/GenerateDocumentation.rst [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+    {{ arg.name }}{% if arg.default %} = '{{ arg.default }}'{% endif %}
+
+.. dropdown:: Returned values data
+
+    Returned values are provided as list of variables names in ``keyword.returns`` variable.
```
Comment (mnojek) [inline docs/source/transformers/GenerateDocumentation.rst:248]:
```suggestion
    Returned values are provided as a list of variables names in ``keyword.returns`` variable.
```

File: tests/atest/transformers/GenerateDocumentation/test_transformer.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    def test_transformer(self):
+        self.compare(source="test.robot")
+
+    def test_transformer_overwrite(self):
+        self.compare(source="test.robot", expected="overwrite.robot", config=":overwrite=True")
+
+    def test_template_with_defaults(self):
+        template_path = Path(__file__).parent / "source" / "template_with_defaults.txt"
+        args = ["--transform", f"{self.TRANSFORMER_NAME};doc_template={template_path}"]
```
Comment (mnojek) [inline tests/atest/transformers/GenerateDocumentation/test_transformer.py:21]:
```suggestion
        args = ["--transform", f"{self.TRANSFORMER_NAME}:doc_template={template_path}"]
```

Comment (mnojek) [reply tests/atest/transformers/GenerateDocumentation/test_transformer.py:21]: That makes me think that there should be one more test to check non-default template also.

File: robotidy/transformers/GenerateDocumentation.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        self.returns = returns
+
+
+class FormattingData:
+    def __init__(self, cont_indent, separator):
+        self.cont_indent = cont_indent
+        self.separator = separator
+
+
+class ArgumentsAndsReturnsVisitor(ModelVisitor):
```
Comment (mnojek) [inline robotidy/transformers/GenerateDocumentation.py:51]:
```suggestion
class ArgumentsAndReturnsVisitor(ModelVisitor):
```

File: robotidy/transformers/GenerateDocumentation.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    """
+
+    ENABLED = False
+
+    WHITESPACE_PATTERN = re.compile(r"(\s{2,}|\t)", re.UNICODE)
+
+    def __init__(self, overwrite: bool = False, doc_template: str = "google", template_directory: Optional[str] = None):
+        self.overwrite = overwrite
+        self.doc_template = Template(self.get_template(doc_template, template_directory))
+        self.args_returns_finder = ArgumentsAndsReturnsVisitor()
```
Comment (mnojek) [inline robotidy/transformers/GenerateDocumentation.py:134]:
```suggestion
        self.args_returns_finder = ArgumentsAndReturnsVisitor()
```

File: robotidy/transformers/GenerateDocumentation.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    ```
+    """
+
+    ENABLED = False
+
+    WHITESPACE_PATTERN = re.compile(r"(\s{2,}|\t)", re.UNICODE)
+
+    def __init__(self, overwrite: bool = False, doc_template: str = "google", template_directory: Optional[str] = None):
+        self.overwrite = overwrite
+        self.doc_template = Template(self.get_template(doc_template, template_directory))
```
Comment (mnojek) [inline robotidy/transformers/GenerateDocumentation.py:133]: I think you need to add an error checking when loading template - if it is correct or not.

File: docs/releasenotes/4.0.0.rst [hunk 7]
```diff
@@ -145,6 +149,37 @@ To be more explicit, this transformer replace such values with ``${EMPTY}`` vari
...
+        [Arguments]    ${arg}
+        ${var}   ${var2}    Step
+        RETURN    ${var}    ${var2}
+
+it will generate::
+
+    *** Keywords ***
+    Keyword
+        [Documentation]
+        ...
```
[the remark this answers is not in the document]
Comment (bhirsz) [reply docs/releasenotes/4.0.0.rst:170]: Good idea

File: robotidy/transformers/GenerateDocumentation.py (hunk 4 above)
Comment (bhirsz) [reply robotidy/transformers/GenerateDocumentation.py:51]: Took me solid 30s to find the difference. Thanks!

File: robotidy/transformers/GenerateDocumentation.py (hunk 6 above)
Comment (bhirsz) [reply robotidy/transformers/GenerateDocumentation.py:133]: Oh, totally forgot about possible errors when rendering jinja template. Thanks, I will add it

File: tests/atest/transformers/GenerateDocumentation/test_transformer.py (hunk 3 above)
Comment (bhirsz) [reply tests/atest/transformers/GenerateDocumentation/test_transformer.py:21]: I have checked - ``;`` is required separator if value contains ``:``. But I will add additional test with relative path and ``:``. I will update the documentation with information on ``;` separator.

File: docs/source/transformers/GenerateDocumentation.rst [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+GenerateDocumentation
+================================
+Generate keyword documentation with the documentation template.
+
+.. |TRANSFORMERNAME| replace:: GenerateDocumentation
+.. include:: disabled_hint.txt
+
+
+By default, GenerateDocumentation uses Google documentation template.
```
[the remark this answers is not in the document]
Comment (bhirsz) [reply docs/source/transformers/GenerateDocumentation.rst]: Added, but now I see that Google documentation uses "Args:" and I've used "Arguments:":. It's better to stick to Google docstring or use more verbose version?

[the remark this answers is not in the document]
Comment (mnojek) [reply docs/source/transformers/GenerateDocumentation.rst]: I would say to stick to the version that Google references. "Args" are commonly understandable, I guess. And they're shorter.

**Review by mnojek — APPROVED** [review]

**Comment by bhirsz** [comment]
Updated with placeholder descriptions

**Review by mnojek — APPROVED** [review]

**Pull request merged**